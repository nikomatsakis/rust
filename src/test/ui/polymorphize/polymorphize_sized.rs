// Simple test for the polymorphize analysis code.
//
// compile-flags: -Zpolymorphize -Zpolymorphize-dump
// compile-pass

#![allow(warnings)]

struct EmbedRef<'a, T: ?Sized> {
    t: &'a T
}

fn no_dependency_because_pointer<T>(t: &T) -> &T {
    //~^ ERROR no polymorphic dependencies found
    t
}

fn dependency_because_unsized_pointer<T: ?Sized>(t: &T) -> &T {
    //~^ ERROR some polymorphic dependencies found
    t
}

fn dependency_because_embed_ref_unsized<T: ?Sized>(t: &T) -> EmbedRef<'_, T> {
    //~^ ERROR some polymorphic dependencies found
    EmbedRef { t }
}

fn dependency_because_embed_ref_sized<T>(t: &T) -> EmbedRef<'_, T> {
    //~^ ERROR no polymorphic dependencies found
    //
    // Here, the size of `EmbedRef` is known up front.
    EmbedRef { t }
}

fn no_dependency_because_unsized_pointer_indirect<T>(t: &T) {
    //~^ ERROR no polymorphic dependencies found

    // Although `dependency_because_unsized_pointer` has a dependency
    // (as T is not known to be sized), *this* function knows that `T`
    // is sized, and that's good enough.
    dependency_because_unsized_pointer::<T>(t);
}

fn main() {
    //~^ ERROR no polymorphic dependencies found

    // Invoke each function so that they are considered by the collector.
    no_dependency_because_pointer::<u32>(&22);
    no_dependency_because_pointer::<u16>(&22);

    dependency_because_unsized_pointer::<u32>(&22);
    dependency_because_unsized_pointer::<u16>(&22);

    dependency_because_embed_ref_unsized::<u32>(&22);
    dependency_because_embed_ref_unsized::<u16>(&22);

    dependency_because_embed_ref_sized::<u32>(&22);
    dependency_because_embed_ref_sized::<u16>(&22);

    no_dependency_because_unsized_pointer_indirect::<u32>(&22);
    no_dependency_because_unsized_pointer_indirect::<u16>(&22);
}
