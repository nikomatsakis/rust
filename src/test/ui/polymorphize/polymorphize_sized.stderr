error: no polymorphic dependencies found
  --> $DIR/polymorphize_sized.rs:11:1
   |
LL | / fn no_dependency_because_pointer<T>(t: &T) -> &T {
LL | |     //~^ ERROR no polymorphic dependencies found
LL | |     t
LL | | }
   | |_^

error: some polymorphic dependencies found
  --> $DIR/polymorphize_sized.rs:16:1
   |
LL | / fn dependency_because_unsized_pointer<T: ?Sized>(t: &T) -> &T {
LL | |     //~^ ERROR some polymorphic dependencies found
LL | |     t
   | |     - depends on `SizeAlignment(&T)`
LL | | }
   | |_^

error: some polymorphic dependencies found
  --> $DIR/polymorphize_sized.rs:21:1
   |
LL | / fn dependency_because_embed_ref_unsized<T: ?Sized>(t: &T) -> EmbedRef<'_, T> {
LL | |     //~^ ERROR some polymorphic dependencies found
LL | |     EmbedRef { t }
   | |     --------------
   | |     |          |
   | |     |          depends on `SizeAlignment(&T)`
   | |     depends on `SizeAlignment(T)`
LL | | }
   | |_^

error: no polymorphic dependencies found
  --> $DIR/polymorphize_sized.rs:26:1
   |
LL | / fn dependency_because_embed_ref_sized<T>(t: &T) -> EmbedRef<'_, T> {
LL | |     //~^ ERROR no polymorphic dependencies found
LL | |     //
LL | |     // Here, the size of `EmbedRef` is known up front.
LL | |     EmbedRef { t }
LL | | }
   | |_^

error: no polymorphic dependencies found
  --> $DIR/polymorphize_sized.rs:33:1
   |
LL | / fn no_dependency_because_unsized_pointer_indirect<T>(t: &T) {
LL | |     //~^ ERROR no polymorphic dependencies found
LL | |
LL | |     // Although `dependency_because_unsized_pointer` has a dependency
...  |
LL | |     dependency_because_unsized_pointer::<T>(t);
   | |     ------------------------------------------ invokes `DefId(0/0:5 ~ polymorphize_sized[317d]::dependency_because_unsized_pointer[0])` with substitutions `[T]`
LL | | }
   | |_^

error: no polymorphic dependencies found
  --> $DIR/polymorphize_sized.rs:42:1
   |
LL | / fn main() {
LL | |     //~^ ERROR no polymorphic dependencies found
LL | | }
   | |_^

error: aborting due to 6 previous errors

